---
title: "Integration of VMH datasets"
author: "Alan Rupp"
date: "`r format(Sys.time(), '%b %d, %Y')`"
output: pdf_document
geometry: margin=0.5in
---

Determining homologous cell populations from our dataset and from Kim et al.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE,
                      fig.align = "center")
library(tidyverse)
library(Seurat)
library(knitr); library(kableExtra)
options(bitmapType = "cairo")
options(future.globals.maxSize = 4000 * 1024^2)
map(list.files("../../../../tools/dropseq3/R", full.names = TRUE), source)
```

# Data
Both datasets were first analyzed separately to determine lab-specific cluster designations.

```{r, include = FALSE}
us <- readRDS("../mouse/vmh.rds")
kim <- readRDS("../kim/vmh.rds")
datasets <- c("us", "kim")
```

```{r}
# overview of samples
data.frame(
  "Dataset" = str_to_title(datasets),
  "Samples" = map_int(datasets, ~ length(unique(get(.x)$Sample))),
  "Neurons" = format(map_int(
    datasets,
    ~ ncol(get(.x)@assays$RNA@counts)
  ), big.mark = ","),
  "Clusters" = map_int(
    datasets,
    ~ length(unique(get(.x)@active.ident))
  )
) %>%
  kable(caption = "Overview") %>%
  kable_styling(latex_options = "hold_position")
```

```{r}
# set Kim clusters to their class designations
kim_levels <- c(
  paste("Dlk1", seq(6), sep = "_"),
  paste("Esr1", c(1, "2|3", 4:7), sep = "_"),
  "Nup62cl",
  paste("Satb2", 1:3, sep = "_"),
  "Scgn", "Tsix_Esr1",
  "(Missing)",
  paste("Nr5a1_Foxp2", 1:2, sep = "_"),
  paste("Nr5a1", c(seq(3), "4|5", 6:8, "9|10", 11), sep = "_"),
  "Doublets"
)

kim@active.ident <- factor(kim$tv_cluster_label)
kim@active.ident <- fct_explicit_na(kim@active.ident)
kim@active.ident <- factor(kim@active.ident, levels = kim_levels)
```

\pagebreak

# Correlation
Which clusters have highly correlated gene expression profiles across datasets?

```{r, fig.height=2, fig.width = 6, include = FALSE}
kim@misc$markers <- FindAllMarkers(kim, only.pos = TRUE, verbose = FALSE)
var_genes <- union(unique(us@misc$markers$gene), unique(kim@misc$markers$gene))

kim <- scale_data(kim, genes = var_genes, group = "Sample")

# keep genes in rownames
var_genes <- var_genes[var_genes %in% rownames(us@assays$RNA@scale.data)]
var_genes <- var_genes[var_genes %in% rownames(kim@assays$RNA@scale.data)]

# Remove var genes with NA scale data
var_genes <- var_genes[!apply(us@assays$RNA@scale.data[var_genes, ], 1, function(x) any(is.na(x)))]
var_genes <- var_genes[!apply(kim@assays$RNA@scale.data[var_genes, ], 1, function(x) any(is.na(x)))]

correlation <- expand.grid(unique(us@active.ident), unique(kim@active.ident))
correlation$corr <- apply(correlation, 1, function(x) {
  cor(
    Matrix::rowMeans(us@assays$RNA@scale.data[
      var_genes, names(us@active.ident)[us@active.ident == x["Var1"]]
      ]),
    Matrix::rowMeans(kim@assays$RNA@scale.data[
      var_genes, names(kim@active.ident)[kim@active.ident == x["Var2"]]
      ])
    )
})

correlation %>%
  ggplot(aes(x = Var1, y = fct_rev(Var2), fill = corr)) +
  geom_tile() +
  scale_fill_gradient2() +
  theme_void() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y = element_text(hjust = 1, vjust = 0.5))

write.csv(correlation, "cluster_correlation.csv", row.names = FALSE)
```

```{r}
# assign colors based on correlation with our clusters
broad_corr <- pivot_wider(correlation, names_from = "Var1", values_from = "corr")
broad_corr <- as.data.frame(broad_corr) %>% column_to_rownames("Var2")
broad_corr <- apply(broad_corr, c(1,2), function(x) ifelse(x < 0, 0, x))
broad_corr <- broad_corr[, levels(us@active.ident)]

# grab our colors
rgb_colors <- col2rgb(us@misc$cluster_colors)
rgb_colors <- rgb_colors[, levels(us@active.ident)]

rgb2hex <- function(v) rgb(v[1], v[2], v[3], maxColorValue = 255)
cluster_colors <- apply(broad_corr, 1, function(x) {
  total = sum(x)
  rgb_color = sapply(1:length(x), function(i) x[i]/total * rgb_colors[, i])
  rgb2hex(rowSums(rgb_color))
})

kim@misc$cluster_colors <- cluster_colors
saveRDS(kim, "../kim/vmh.rds")
```

```{r, fig.height = 3, fig.width = 6}
plot_settings <- list(
  theme_void(),
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5))
)
cowplot::plot_grid(plotlist = list(
  DimPlot(us, label = TRUE) +
    scale_color_manual(values = us@misc$cluster_colors) +
    ggtitle("Affinati et al.") +
    plot_settings,
  DimPlot(kim, label = TRUE) +
    scale_color_manual(values = kim@misc$cluster_colors) +
    ggtitle("Kim et al.") +
    plot_settings
))
```

\pagebreak

# Integration
Integrating data using the Seurat CCA framework.

```{r}
# keep metadata for future use
metadata <- map(
  datasets,
  ~ data.frame(
    "cell" = rownames(get(.x)@meta.data),
    "Sample" = get(.x)$Sample,
    "Cluster" = get(.x)@active.ident)
)
names(metadata) <- datasets
```

```{r}
# grab just the counts slot from each
srt <- map(datasets, ~ get(.x)@assays$RNA@counts)
```

```{r, include = FALSE}
# normalize data and find features
srt <- map(srt, CreateSeuratObject)
srt <- map(srt, normalize_data)
srt <- map(srt, FindVariableFeatures)
gc()
```

Now running the Seurat CCA pipeline ...

```{r}
srt <- integrate_data(srt)
```

```{r, include = FALSE}
srt <- ScaleData(srt, vars.to.regress = "nFeature_RNA")
srt <- RunPCA(srt, npcs = 100)
srt@reductions$pca@misc$sig_pcs <- knee_test(srt)
srt <- RunUMAP(srt, dims = 1:srt@reductions$pca@misc$sig_pcs)
```

Looking for dataset-specific and sequencing depth effects ...

```{r, fig.height = 3.5, fig.width = 7.5}
srt@meta.data <- cbind(srt@meta.data, do.call(rbind, metadata))
srt$Dataset <- rep(datasets, times = map_int(datasets, ~ ncol(get(.x)))) %>%
    str_to_title()

cowplot::plot_grid(plotlist = list(
  DimPlot(srt, group.by = "Dataset", cells = sample(Cells(srt), 10000)) +
    theme_void() +
    ggtitle("Dataset"),
  FeaturePlot(srt, "nFeature_RNA", cells = sample(Cells(srt), 10000)) +
    theme_void() +
    ggtitle("Genes / cell") +
    scale_color_viridis_c(option = "magma")
))

```

```{r, include = FALSE}
gc()
```

This looks great. Continuing with clustering.

\pagebreak

# Clustering integrated data

To cluster, I'll start with standard Louvain clustering from the kNN graph. With `k` set at 20, I'll vary the `resolution` parameter to optimize mean silhouette width.

```{r}
resolutions <- seq(0.2, 2, by = 0.2)
print(paste("Finding clusters for resolutions", paste(resolutions, collapse = ", "), "..."))
```

```{r, include = FALSE}
sig_pcs <- srt@reductions$pca@misc$sig_pcs
srt <- FindNeighbors(srt, dims = 1:sig_pcs)

clusters <- map_dfc(
  resolutions,
  ~ FindClusters(srt, resolution = .x)@active.ident
)
gc()
# get dissimilarity matrix
cells <- sample(Cells(srt), 10000)
distances <- cluster::daisy(
  srt@reductions$pca@cell.embeddings[cells, 1:sig_pcs]
)
gc()

# calculate silhouette width for each clustering
mean_widths <- apply(clusters[match(cells, Cells(srt)), ], 2, function(x) {
  silhouette_object <- cluster::silhouette(as.numeric(x), distances)
  summary(silhouette_object)[["avg.width"]]
})

# choose best resolution based on max mean silhoutte width
best_resolution <- resolutions[mean_widths == max(mean_widths)]

# add clustering to official ident slot
srt@active.ident <- clusters[, resolutions == best_resolution][[1]]

# remove unnecessary object
rm(resolutions, clusters, sig_pcs, distances, mean_widths); gc()
```

Using resolution `r best_resolution`, there are `r length(unique(srt@active.ident))` populations.

```{r}
# order clusters
cluster_order <- order_clusters(srt, srt[["integrated"]]@var.features,
                                assay = "integrated")
srt@active.ident <- factor(
  srt@active.ident,
  levels = cluster_order,
  labels = 1:length(cluster_order)
)
```

```{r}
cluster_names <- c(
  `15` = "Dlk1_1",
  `16` = "Dlk1_2",
  `1` = "Esr1_1",
  `3` = "Esr1_2",
  `5` = "Esr1_3",
  `2` = "Nup62cl",
  `4` = "Satb2",
  `12` = "Nfib",
  `13` = "Doublets_1",
  `14` = "Doublets_2",
  `6` = "Foxp2_1",
  `7` = "Foxp2_2",
  `10` = "Fezf1_1",
  `11` = "Fezf1_2",
  `8` = "Lepr_1",
  `9` = "Lepr_2"
)

srt$Merged_cluster <- srt@active.ident
srt@active.ident <- recode(srt@active.ident, !!!cluster_names)
srt@active.ident <- factor(srt@active.ident, levels = cluster_names)
```

```{r}
# choose colors based on our population colors
class_table <- table(
  srt@active.ident[srt$Dataset == "Us"],
  srt$Cluster[srt$Dataset == "Us"]
)
class_table <- class_table[, names(us@misc$cluster_colors)]
rgb_colors <- col2rgb(us@misc$cluster_colors)
rgb2hex <- function(v) rgb(v[1], v[2], v[3], maxColorValue = 255)
cluster_colors <- apply(class_table, 1, function(x) {
  total = sum(x)
  rgb_color = sapply(1:length(x), function(i) x[i]/total * rgb_colors[, i])
  rgb2hex(rowSums(rgb_color))
})
names(cluster_colors) <- recode(names(cluster_colors), !!!cluster_names)
srt@misc$cluster_colors <- cluster_colors
```

\pagebreak

# Comparison
Comparing integrated clusters to original clusters

```{r, fig.width = 6, fig.height = 3}
srt$original <- c(
  as.character(us@active.ident),
  as.character(kim@active.ident)
)

plot_new <- function(dataset, cluster_levels, colors) {
  cells <- colnames(srt)[srt$Dataset == dataset]
  df <- data.frame(
    "UMAP1" = srt@reductions$umap@cell.embeddings[cells, 1],
    "UMAP2" = srt@reductions$umap@cell.embeddings[cells, 2],
    "cluster" = srt$original[cells]
  ) %>%
    mutate(cluster = factor(cluster, levels = cluster_levels))
  cluster_labels <- df %>% group_by(cluster) %>%
    summarize("UMAP1" = median(UMAP1), "UMAP2" = median(UMAP2))
  p <- ggplot(df, aes(x = UMAP1, y = UMAP2)) +
    geom_point(stroke = 0, aes(color = cluster), show.legend = FALSE,
               size = 1) +
    scale_color_manual(values = colors) +
    geom_text(data = cluster_labels, aes(label = cluster)) +
    theme_void() +
    theme(plot.title = element_text(hjust = 0.5))
  return(p)
}

cowplot::plot_grid(plotlist = list(
   plot_new("Us", levels(us@active.ident), us@misc$cluster_colors) +
     ggtitle("Affinati, Sabatini et al."),
  plot_new("Kim", levels(kim@active.ident), kim@misc$cluster_colors) +
     ggtitle("Kim et al.")
))
```


```{r, fig.width = 6, fig.height = 5.5}
cowplot::plot_grid(plotlist = list(
  data.frame("cluster" = srt@active.ident[srt$Dataset == "Us"],
             "original" = srt$original[srt$Dataset == "Us"]) %>%
    mutate(original = fct_explicit_na(original)) %>%
    mutate(original = factor(original, levels = levels(us@active.ident))) %>%
  group_by(cluster, original) %>%
  count() %>%
  ggplot(aes(x = fct_rev(cluster), y = -n, fill = original)) +
  geom_col(color = "black") +
    scale_x_discrete(drop = FALSE) +
    scale_fill_manual(values = us@misc$cluster_colors) +
  coord_flip() +
  theme_void() +
    theme(legend.position = "left",
          plot.title = element_text(hjust = 0.5)) +
    guides(fill = guide_legend(ncol = 1)) +
    ggtitle("Affinati, Sabatini et al."),
  data.frame("cluster" = srt@active.ident[srt$Dataset == "Kim"],
             "original" = srt$original[srt$Dataset == "Kim"]) %>%
    mutate(original = fct_explicit_na(original)) %>%
    mutate(original = factor(original, levels = levels(kim@active.ident))) %>%
  group_by(cluster, original) %>%
  count() %>%
  ggplot(aes(x = fct_rev(cluster), y = n, fill = original)) +
  geom_col(color = "black") +
    scale_fill_manual(values = kim@misc$cluster_colors) +
    guides(fill = guide_legend(ncol = 1)) +
  coord_flip() +
  theme_void() +
  theme(axis.text.y = element_text(),
        legend.position = "right",
        plot.title = element_text(hjust = 0.5),
        legend.key.height = unit(0.15, "in"),
        legend.key.width = unit(0.15, "in")) +
    ggtitle("Kim et al.")
), rel_widths = c(0.45, 0.55))
```

\pagebreak

# Files
Saving Seurat object as `vmh.rds`

```{r}
saveRDS(srt, "vmh.rds")
```

\pagebreak

# Session info
```{r}
sessionInfo()
```
