---
title: "Macaque VMH"
author: "Alan Rupp"
date: "`r format(Sys.time(), '%b %d, %Y')`"
output: pdf_document
geometry: margin=0.5in
---

Analyzing macaque hypothalamus data from the VMH and DMH

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE,
                      fig.align = "center")
library(tidyverse)
library(Seurat)
library(knitr); library(kableExtra)
options(bitmapType = "cairo")
options(future.globals.maxSize = 3000 * 1024^2)
map(list.files("../../../../tools/dropseq3/R", full.names = TRUE), source)
```

Loading data ...

```{r read}
# read in all data
dirs <- list.dirs(path = '../../data')
dirs <- dirs[str_detect(dirs, "NovaA-225")]
dirs <- dirs[str_detect(dirs, "filtered")]
samples <- str_extract(dirs, "(?<=[0-9]\\/)(.+?)(?=\\/outs)")

# read in data as a list of matrices
mtx <- map(dirs, Read10X)
```

There are `r length(samples)` samples and `r format(sum(map_int(mtx, ncol)), big.mark = ",")` total cells.

```{r}
# table of samples and total cells
data.frame(
  "Sample" = samples,
  "Nuclei" = format(map_int(mtx, ncol), big.mark = ",")
  ) %>%
  kable(caption = "Samples") %>%
  kable_styling(full_width = FALSE, position = "left",
                latex_options = "hold_position")
```

\pagebreak

# Quality control
## Keeping high quality genes
I'm keeping only genes expressed in at least 4 cells and present in all `r length(mtx)` samples.

```{r genes-filtering}
print(paste("Starting with", max(map_int(mtx, nrow)), "genes ..."))
mtx <- keep_shared_genes(mtx)
mtx <- remove_low_abundance_genes(mtx)
print(paste("Now", max(map_int(mtx, nrow)), "genes."))
```

## Keeping high quality cells
I'm going to remove any cells that have fewer than 500 genes detected.

```{r filter-cells}
print(paste("Starting with", sum(map_int(mtx, ncol)), "cells ..."))
mtx <- filter_cells(mtx)
mtx <- rename_duplicates(mtx)
print(paste("Now", sum(map_int(mtx, ncol)), "cells."))
```

Continuing with data analysis.

---

```{r create-seurat, include = FALSE}
# combine into Seurat object
srt <- CreateSeuratObject(do.call(cbind, mtx))

# add metadata
srt$Sample <- rep(samples, times = map_dbl(mtx, ncol))
rm(mtx, samples)
gc()
```

```{r metadata-colors}
run_colors <- rcartocolor::carto_pal(3, "Antique")[1]

sample_colors <- rcartocolor::carto_pal(3, "Antique")[2:3]
sample_colors <- order_colors(sample_colors)
names(sample_colors) <- sort(unique(srt$Sample))

# stash in object for later retrieval
srt@misc$run_colors <- run_colors
srt@misc$sample_colors <- sample_colors
```

# Quality control
```{r, fig.width = 2.2, fig.height = 1.9}
fig_cells <- table(srt$Sample) %>%
  as.data.frame() %>%
  filter(Freq != 0) %>%
  rename("Sample" = Var1) %>%
  select(-starts_with("Var")) %>%
  ggplot(aes(x = Sample, y = Freq)) +
  geom_col(color = "black", fill = srt@misc$run_colors) +
  ggtitle("Cells") +
  theme_void() +
  theme(axis.text.x = element_text(color = "black", angle = 90,
                                   hjust = 1, vjust = 0.5, size = 8),
        axis.text.y = element_text(color = "black", size = 7),
        plot.title = element_text(size = 10, hjust = 0.5),
        legend.key.width = unit(0.1, "in"),
        legend.key.height = unit(0.05, "in"))

```

```{r, fig.width = 4, fig.height = 1.9}
plot_metadata <- function(measure, title = NULL) {
  data.frame(
    "metadata" = srt@meta.data[, measure],
    "Sample" = srt$Sample
    ) %>%
    ggplot(aes(x = Sample, y = metadata)) +
    geom_violin(fill = srt@misc$run_colors) +
    scale_y_continuous(trans = "log2") +
    ggtitle(title) +
    theme_void() +
  theme(axis.text.x = element_text(color = "black", angle = 90,
                                   hjust = 1, vjust = 0.5, size = 8),
        axis.text.y = element_text(color = "black", size = 7),
        strip.text = element_text(size = 10),
        legend.title = element_text(size = 9),
        legend.key.height = unit(0.02, "in"),
        legend.key.width = unit(0.08, "in"),
        plot.title = element_text(hjust = 0.5, size = 10))
}

fig_genes <- plot_metadata("nFeature_RNA", "Genes") +
  theme(legend.position = "none")
fig_umis <- plot_metadata("nCount_RNA", "UMIs")
```

```{r plot-metadata, fig.height = 1.9, fig.width = 4.5}
cowplot::plot_grid(fig_cells, fig_genes, fig_umis, ncol = 3)
```

Good gene and UMI depth for both samples.


\pagebreak

# Identifying cell types
My main goal is simply to identify neurons for downstream analysis, but I'm going to predict cell type for each cell and cluster all cells so we'll have the data in case we want to look at glia.

Pipeline:
1. Normalizing data with `scran` `r packageVersion("scran")` for each run independently
2. Finding variable genes with `Seurat`
3. Scaling variable genes for each sample independently
4. Running PCA on scaled data
5. Choosing PCs from the elbow on the scree plot
6. Projecting into UMAP space

```{r process-all, include = FALSE}
srt <- normalize_data(srt, method = "scran")
srt <- FindVariableFeatures(srt)
srt <- scale_data(srt, genes = srt@assays$RNA@var.features, groups = "Sample")
srt <- RunPCA(srt, npcs = 100)
srt@reductions$pca@misc$sig_pcs <- knee_test(srt)
srt <- RunUMAP(srt, dims = 1:srt@reductions$pca@misc$sig_pcs)
gc()
```

Results: using `r length(srt@assays$RNA@var.features)` variable genes and `r srt@reductions$pca@misc$sig_pcs` PCs.

```{r dimplot-sample-all, fig.height = 3, fig.width = 4}
DimPlot(srt, group.by = "Sample", cells = sample(Cells(srt))) +
  scale_color_manual(values = srt@misc$sample_colors) +
  theme_void() +
  theme(legend.text = element_text(size = 7),
        plot.title = element_blank())
```

This looks great, continuing with identifying cell types.


\pagebreak

# Predicting cell types
I'm going to use published scRNA-seq data (from Campbell et al. and Chen et al.) to infer cell types in our dataset. In short, I'll use the top marker genes from each celltype in those dataset to assign a score to each cell in our dataset. Then, I'll cluster our cells and assign celltype membership based on the score, plus whether they are a presumptive doublet (from the `scrublet` package). I'll refer to doublets & unclassifed cells as NA.

```{r}
print("Clustering ...")
```

```{r cluster-all, include = FALSE}
srt <- FindNeighbors(srt, dims = 1:srt@reductions$pca@misc$sig_pcs)
srt <- FindClusters(srt)
```

There are `r length(unique(srt@active.ident))` clusters.

```{r, include = FALSE}
# find cluster markers
cluster_order <- order_clusters(srt, srt@assays$RNA@var.features)
srt@active.ident <- factor(
  srt@active.ident,
  levels = cluster_order,
  labels = 1:length(cluster_order)
)
srt$seurat_clusters <- srt@active.ident
```

```{r}
color_options <- map(
  names(wesanderson::wes_palettes),
  ~ as.character(wesanderson::wes_palette(.x))
) %>%
  unlist()

# only keep colors that aren't too light or dark
get_luminance <- function(color) {
  col <- col2rgb(color)
  0.2126*col["red", ] + 0.7152*col["green", ] + 0.0722*col["blue" ,]
}

# cluster similar colors for each class
color_options <- color_options[
  get_luminance(color_options) > 100 & get_luminance(color_options) < 200
]

# skip along colors to get full coverage
color_options <- order_colors(color_options)
skip <- length(color_options)/length(levels(srt@active.ident))
cluster_colors <- color_options[
  round(seq(skip, skip*length(levels(srt@active.ident)), by = skip), 0)
]
names(cluster_colors) <- levels(srt@active.ident)
srt@misc$cluster_colors <- cluster_colors
```

```{r, fig.width = 2.5, fig.height = 2.5}
dim_plot(srt, label_size = 2, point_size = 0.4) +
  scale_color_manual(values = srt@misc$cluster_colors)
```


```{r find-celltypes}
print("Predicting cell type membership ...")
markers <- read_csv("../celltypes/markers.csv")
# turn genes into macaque orthologs
orthologs <- read_tsv("../../data/mouse_to_macaque.tsv.gz")
markers$Gene <- orthologs$`Gene name_1`[
  match(markers$Gene, orthologs$`Gene name`)
]
# only keep markers in the Seurat object
markers <- filter(markers, !is.na(Gene))
markers <- filter(markers, Gene %in% rownames(srt))
markers <- filter(markers, !duplicated(Gene))
markers <- as.data.frame(markers) %>% column_to_rownames("Gene")

# get top 100 markers for each cell type
top_markers <- apply(markers, 2, function(x) {
  names(x) = rownames(markers)
  sort(x, decreasing = TRUE)[1:100] %>% names()
})

# scale data for top markers
srt <- scale_data(srt, genes = c(top_markers), groups = "Sample"); gc()

if (ncol(srt) > 10000) {
  cells <- sample(Cells(srt), 10000)
} else {
  cells <- Cells(srt)
}
scaled <- srt@assays$RNA@scale.data[, cells]

# predict celltype by max cumsum for scaled marker genes
get_celltype <- function(cell) {
  scores <- sapply(colnames(top_markers), function(x) {
    max(cumsum(scaled[top_markers[, x], cell]), na.rm = TRUE)  
  })
  if (max(scores, na.rm = TRUE) < 0) {
    "Unclassified"
  } else {
    names(scores)[scores == max(scores)]
  }
}
celltypes <- sapply(cells, get_celltype)

# call celltype by cluster
cluster_table <- table(srt@active.ident[cells], celltypes)
cluster_celltypes <- apply(cluster_table, 1, function(x) {
  if (any(x > sum(x)/2)) {
    names(x)[x == max(x)]
  } else {
    "Mixed"
  }
})
srt$Celltype <- recode(srt@active.ident, !!!cluster_celltypes)
```

```{r}
table(srt$Celltype) %>%
  as.data.frame() %>%
  mutate("Percent" = round(Freq / sum(Freq) * 100, 1)) %>%
  arrange(desc(Percent)) %>%
  rename("Cells" = Var1, "N" = Freq) %>%
  kable(caption = "Cell type breakdown") %>%
  kable_styling(latex_options = "hold_position")
```

```{r, fig.width = 6, fig.height = 3}
srt@active.ident <- factor(srt$Celltype)
cluster_order <- order_clusters(srt, unique(c(top_markers)))
srt@active.ident <- factor(srt@active.ident, levels = cluster_order)

celltypes <- srt$Celltype
names(celltypes) <- srt@active.ident
celltypes <- celltypes[!duplicated(names(celltypes))]
celltypes <- celltypes[order(factor(names(celltypes), levels = 1:length(celltypes)))]

rgb2hex <- function(r, g, b) rgb(r, g, b, maxColorValue = 255)
average_colors <- function(colors, groupby) {
  groups <- unique(groupby)
  sapply(groups, function(x) {
    mtx <- col2rgb(colors[groupby == x])
    avg <- rowMeans(mtx)
    rgb2hex(avg["red"], avg["green"], avg["blue"])
  })
}
class_colors <- average_colors(srt@misc$cluster_colors, celltypes)
names(class_colors) <- unique(celltypes)
srt@misc$celltype_colors <- class_colors

cowplot::plot_grid(plotlist = list(
  DimPlot(srt, label = TRUE, cells = sample(Cells(srt), 10000),
          group.by = "Celltype") +
    scale_color_manual(values = srt@misc$celltype_colors) +
    theme_void() +
    theme(legend.position = "none"),
  table(srt$Sample, srt$Celltype) %>%
    as.data.frame() %>%
    mutate(Var1 = factor(Var1, levels = unique(srt$Sample))) %>%
    group_by(Var1) %>%
    mutate("Percent" = Freq / sum(Freq) * 100) %>%
    ggplot(aes(x = Var1, y = Percent, fill = Var2)) +
    geom_col() +
    xlab(NULL) +
    ylab("% of dataset") +
    scale_y_continuous(expand = c(0, 0)) +
    scale_fill_manual(values = srt@misc$celltype_colors, name = NULL) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5,
                                     color = "black", size = 8),
          axis.text.y = element_text(size = 7),
          axis.title.y = element_text(size = 8),
          legend.key.width = unit(0.08, "in"),
          legend.key.height = unit(0.04, "in"),
          legend.text = element_text(size = 7))
))
```

Good balance of cell types across all samples.

---

I'm saving the all-cell object as `macaque.rds` and I'll move on to just neuron analysis.

```{r}
saveRDS(srt, "macaque.rds")
```

\pagebreak

# Neurons
There are `r format(sum(srt$Celltype == "Neuron"), big.mark = ",")` neurons. I'll process them separately now to identify neuronal populations.

```{r, include = FALSE}
neurons <- subset(srt, Celltype == "Neuron")
rm(srt); gc()
```

Pipeline:
1. Finding variable genes with `Seurat`
2. Running `sctransform` with Run as a batch variable.
2. Scaling variable genes (in both RNA and SCT assays) for each sample independently
3. Running PCA on scaled data
4. Choosing PCs from the elbow on the scree plot
5. Projecting into UMAP space

```{r, include = FALSE}
neurons <- FindVariableFeatures(neurons)
neurons <- SCTransform(neurons, conserve.memory = TRUE)
var_genes <- intersect(
  neurons@assays$RNA@var.features, neurons@assays$SCT@var.features
)

# compare RNA and SCT assays for batch effects
DefaultAssay(neurons) <- "RNA"
neurons <- scale_data(neurons, genes = var_genes, groups = "Sample")
neurons <- RunPCA(neurons, npcs = 100)
neurons@reductions$pca@misc$sig_pcs <- knee_test(neurons)
neurons <- RunUMAP(neurons, dims = 1:neurons@reductions$pca@misc$sig_pcs)

umap_object <- neurons[["umap"]]
pca_object <- neurons[["pca"]]

# compare RNA and SCT assays for batch effects
DefaultAssay(neurons) <- "SCT"
neurons <- RunPCA(neurons, npcs = 100)
neurons@reductions$pca@misc$sig_pcs <- knee_test(neurons)
neurons <- RunUMAP(neurons, dims = 1:neurons@reductions$pca@misc$sig_pcs)

gc()
```

Results: Using `r length(neurons@assays$RNA@var.features)` variable genes from `Seurat` found `r length(neurons@assays$SCT@var.features)` from `sctransform`.  The `Seurat` variable features resulted in `r pca_object@misc$sig_pcs` PCs, and the `sctransform` variable genes resulted in `r neurons@reductions$pca@misc$sig_pcs` PCs.

```{r, fig.width = 7, fig.height = 3.2}
plot_umap_samples <- function(umap) {
  df <- bind_cols(as.data.frame(umap), neurons$Sample)
  ggplot(df, aes(x = UMAP_1, y = UMAP_2, color = `...3`)) +
    geom_point(stroke = 0, size = 1) +
    scale_color_manual(values = neurons@misc$sample_colors, name = NULL) +
    theme_void() +
    theme(legend.text = element_text(size = 7),
          plot.title = element_text(size = 10, hjust = 0.5))
}

cowplot::plot_grid(plotlist = list(
  plot_umap_samples(umap_object@cell.embeddings) + ggtitle("RNA"),
  plot_umap_samples(neurons@reductions$umap@cell.embeddings) + ggtitle("SCT")
))
```

The **RNA** assay looks better, cells are more "clumped", even across samples, making it more likely to find clusters.

```{r, include = FALSE}
DefaultAssay(neurons) <- "RNA"
neurons[["pca"]] <- pca_object
neurons[["umap"]] <- umap_object
rm(umap_object, pca_object)
gc()
```

\pagebreak

# Removing doublets

Sometimes neuron/glia doublets make it through this filtering step. To detect doublets, I'm going to run WGCNA on the genes that are highly loaded (>2 SD) onto any of the PCs I used for dimension reduction. Any cells that load highly on gene modules that are overrepresented for glia classes will be removed.

```{r wgcna, include = FALSE}
print("Running WGCNA ...")
# choose variable genes
sig_pcs <- neurons@reductions$pca@misc$sig_pcs
var_genes <- apply(
  neurons@reductions$pca@feature.loadings[, 1:sig_pcs], 2, function(x) {
    scaled = scale(x)
    rownames(neurons@reductions$pca@feature.loadings)[abs(scaled) > 2]
  }
) %>%
  unlist() %>%
  unique()

# find WGCNA modules
library(WGCNA)
sft <- pickSoftThreshold(
  Matrix::t(neurons@assays$RNA@data[var_genes, ]),
  networkType = "signed"
)
softPower <- sft$fitIndices$Power[sft$fitIndices[,"SFT.R.sq"] > 0.8][1]

net <- blockwiseModules(
  Matrix::t(neurons@assays$RNA@data[var_genes, ]),
  power = softPower
)
```

WGCNA identifies `r length(unique(net$colors))` modules from the `r length(var_genes)` genes, using a soft power threshold of `r softPower`.

```{r}
print("Calculating celltype enrichment of each module with hypergeometric test (with BH adjustment) ...")
# find enrichment of modules by celltype
genes_in_both <- intersect(unique(c(top_markers)), var_genes)
enrichment_test <- function(module) {
  module_genes <- intersect(var_genes[net$colors == module], genes_in_both)
  p <- apply(top_markers, 2, function(x) {
    phyper(q = length(intersect(module_genes, x)),
           m = length(genes_in_both),
           n = length(x),
           k = length(module_genes),
           lower.tail = FALSE)
    })
  p.adjust(p, method = "BH")
}
modules <- unique(net$colors)
enrichment <- sapply(modules, enrichment_test)

celltype_modules <- apply(enrichment, 2, function(x) {
  if (any(x < 0.05)) {
    paste(names(x)[x < 0.05], collapse = ",")
  } else {
    NULL
  }
}) %>%
  unlist()
```

`r sum(enrichment < 0.05)` modules are associated with specific cell types: `r paste(map_chr(1:length(celltype_modules), ~ paste(names(celltype_modules)[.x], "=", celltype_modules[.x])), collapse = ", ")`

I'll now remove any cells strongly loaded onto these modules.

```{r, fig.width = 6, fig.height = 3.2}
eigengenes <- map_dfc(
  names(celltype_modules),
  ~ eigengene(neurons, var_genes[net$colors == .x])
) %>%
  set_names(names(celltype_modules))
scaled <- scale(eigengenes)

plot_eigengenes <- function(module) {
  df <- bind_cols(as.data.frame(neurons@reductions$umap@cell.embeddings),
                  data.frame("mod" = scaled[, module]))
  ggplot(df, aes(x = UMAP_1, y = UMAP_2, color = mod > 2)) +
    geom_point(show.legend = FALSE, stroke = 0, size = 0.4) +
    scale_color_manual(values = wesanderson::wes_palette("Royal1")[1:2]) +
    ggtitle(paste0(celltype_modules[module], " module: ",
                   sum(df$mod > 2), " cells")) +
    theme_void() +
    theme(plot.title = element_text(size = 10, hjust = 0.5))
}

cowplot::plot_grid(plotlist = map(names(celltype_modules), plot_eigengenes))
```

```{r}
doublets <- apply(scaled, 1, function(x) any(x > 2))
doublet_cells <- Cells(neurons)[doublets]
singlet_cells <- Cells(neurons)[!doublets]
```

I'll remove these `r sum(doublets)` cells from the dataset as doublets. That leaves `r length(singlet_cells)` neurons. I'll also save the doublets as a Seurat object called `neuron_doublets.rds` in case I want to look at them.

```{r}
saveRDS(subset(neurons, cells = doublet_cells), "neuron_doublets.rds")
neurons <- subset(neurons, cells = singlet_cells)
```

```{r, include = FALSE}
rm(sft, softPower, net, genes_in_both, enrichment_test, modules, enrichment,
   celltype_modules)
gc()
```

\pagebreak
# Pre-processing neurons (again)
After removing doublets, I'll re-process.

Pipeline:
1. Finding variable genes with `Seurat`
2. Running `sctransform` with Run as a batch variable.
2. Scaling variable genes (in both RNA and SCT assays) for each sample independently
3. Running PCA on scaled data
4. Choosing PCs from the elbow on the scree plot
5. Projecting into UMAP space

```{r, include = FALSE}
neurons <- FindVariableFeatures(neurons)
neurons <- SCTransform(neurons, conserve.memory = TRUE)
var_genes <- intersect(
  neurons@assays$RNA@var.features, neurons@assays$SCT@var.features
)

# compare RNA and SCT assays for batch effects
DefaultAssay(neurons) <- "RNA"
neurons <- scale_data(neurons, genes = var_genes, groups = "Sample")
neurons <- RunPCA(neurons, npcs = 100)
neurons@reductions$pca@misc$sig_pcs <- knee_test(neurons)
neurons <- RunUMAP(neurons, dims = 1:neurons@reductions$pca@misc$sig_pcs)

umap_object <- neurons[["umap"]]
pca_object <- neurons[["pca"]]

# compare RNA and SCT assays for batch effects
DefaultAssay(neurons) <- "SCT"
neurons <- RunPCA(neurons, npcs = 100)
neurons@reductions$pca@misc$sig_pcs <- knee_test(neurons)
neurons <- RunUMAP(neurons, dims = 1:neurons@reductions$pca@misc$sig_pcs)

gc()
```

Results: Using `r length(neurons@assays$RNA@var.features)` variable genes from `Seurat` found `r length(neurons@assays$SCT@var.features)` from `sctransform`.  The `Seurat` variable features resulted in `r pca_object@misc$sig_pcs` PCs, and the `sctransform` variable genes resulted in `r neurons@reductions$pca@misc$sig_pcs` PCs.

```{r, fig.width = 7, fig.height = 3.2}
plot_umap_samples <- function(umap) {
  df <- bind_cols(as.data.frame(umap), neurons$Sample)
  ggplot(df, aes(x = UMAP_1, y = UMAP_2, color = `...3`)) +
    geom_point(stroke = 0, size = 1) +
    scale_color_manual(values = neurons@misc$sample_colors, name = NULL) +
    theme_void() +
    theme(legend.text = element_text(size = 7),
          plot.title = element_text(size = 10, hjust = 0.5))
}

cowplot::plot_grid(plotlist = list(
  plot_umap_samples(umap_object@cell.embeddings) + ggtitle("RNA"),
  plot_umap_samples(neurons@reductions$umap@cell.embeddings) + ggtitle("SCT")
))
```

The **RNA** assay looks better, cells are more "clumped", even across samples, making it more likely to find clusters.

```{r, include = FALSE}
DefaultAssay(neurons) <- "RNA"
neurons[["pca"]] <- pca_object
neurons[["umap"]] <- umap_object
rm(umap_object, pca_object)
gc()
```

\pagebreak

# Clustering neurons

To cluster, I'll start with standard Louvain clustering from the kNN graph. With `k` set at 20, I'll vary the `resolution` parameter to optimize mean silhouette width.

```{r}
resolutions <- seq(0.2, 2, by = 0.2)
print(paste("Finding clusters for resolutions", paste(resolutions, collapse = ", "), "..."))
```

```{r, include = FALSE}
sig_pcs <- neurons@reductions$pca@misc$sig_pcs
neurons <- FindNeighbors(neurons, dims = 1:sig_pcs)

clusters <- map_dfc(
  resolutions,
  ~ FindClusters(neurons, resolution = .x)@active.ident
)
gc()
# get dissimilarity matrix
sig_pcs <- neurons@reductions$pca@misc$sig_pcs
distances <- cluster::daisy(
  neurons@reductions$pca@cell.embeddings[, 1:sig_pcs]
)
gc()

# calculate silhouette width for each clustering
mean_widths <- apply(clusters, 2, function(x) {
  silhouette_object <- cluster::silhouette(as.numeric(x), distances)
  summary(silhouette_object)[["avg.width"]]
})

# choose best resolution based on max mean silhoutte width
best_resolution <- resolutions[mean_widths == max(mean_widths)]

# add clustering to official ident slot
neurons@active.ident <- clusters[, resolutions == best_resolution][[1]]

# remove unnecessary object
rm(resolutions, clusters, sig_pcs, distances, mean_widths); gc()
```

Using resolution `r best_resolution`, there are `r length(unique(neurons@active.ident))` populations.

```{r}
cluster_order <- order_clusters(neurons, var_genes)
neurons@active.ident <- factor(
  neurons@active.ident,
  levels = cluster_order,
  labels = 1:length(cluster_order)
)
```

```{r, fig.width = 3, fig.height = 3}
wes_colors <- sapply(names(wesanderson::wes_palettes), wesanderson::wes_palette)
wes_colors <- unlist(wes_colors)
wes_colors <- wes_colors[!duplicated(wes_colors)]

get_luminance <- function(color) {
  col <- col2rgb(color)
  0.2126*col["red", ] + 0.7152*col["green", ] + 0.0722*col["blue" ,]
}
wes_colors <- wes_colors[get_luminance(wes_colors) > 80 &
                         get_luminance(wes_colors) < 220]
cluster_colors <- wes_colors[1:length(levels(neurons@active.ident))]
cluster_colors <- order_colors(cluster_colors)
names(cluster_colors) <- levels(neurons@active.ident)
neurons@misc$cluster_colors <- cluster_colors

DimPlot(neurons, label = TRUE) +
  scale_color_manual(values = neurons@misc$cluster_colors) +
  theme_void() +
  theme(legend.position = "none")
```

\pagebreak

# Identifying VMH neurons
*Fezf1* and *Nr5a1* are VMH marker genes.

```{r, fig.width = 1.5, fig.height = 3}
data.frame(
  "cell" = colnames(neurons),
  "UMAP1" = neurons@reductions$umap@cell.embeddings[,1],
  "UMAP2" = neurons@reductions$umap@cell.embeddings[,2],
  "Nr5a1" = neurons@assays$RNA@data["NR5A1", ],
  "Fezf1" = neurons@assays$RNA@data["FEZF1", ]
) %>%
  gather(-cell, -UMAP1, -UMAP2, key = "gene", value = "expr") %>%
  ggplot(aes(x = UMAP1, y = UMAP2, color = expr)) +
  geom_point(stroke = 0, size = 1) +
  scale_color_gradient(low = "gray90", high = "navyblue",
                       name = "Normalized\nexpression") +
  theme_void() +
  facet_wrap(~gene, ncol = 1) +
  theme(legend.position = "bottom", legend.direction = "horizontal",
        legend.title = element_text(size = 7),
        legend.text = element_text(size = 7),
        legend.key.height = unit(0.03, "in"),
        legend.key.width = unit(0.15, "in"))
```

It seems clear there's a distinct lobe that corresponds to the VMH populations, but I'll confirm this with the *Nr5a1* TRAP-seq.

```{r, fig.width = 1.5, fig.height = 1.5}
print("Reading in Nr5a1-Cre TRAP-seq data ...")
sf1 <- read_csv("../../../TRAP-seq/VMH_Nr5a1/results.csv")
```

I'll use the *Nr5a1-Cre* TRAP-seq enriched genes that are expressed at least 10 CPM and log2 enriched at least 1 to identify VMH neurons in the snRNA-seq.

```{r}
sf1 <- filter(sf1, Bead > 10 & P < 0.05 & Enrichment > 1)
# convert to macaque orthologs
orthologs <- read_tsv("../../data/mouse_to_macaque.tsv.gz")
sf1$Gene <- orthologs$`Gene name_1`[match(sf1$Gene, orthologs$`Gene name`)]
sf1 <- filter(sf1, Gene %in% rownames(neurons))
```

That corresponds to `r nrow(sf1)` genes. I'll use PCA to reduce these enriched genes into a single score for each cell.

```{r}
sf1_score <- eigengene(neurons, sf1$Gene)
neurons$Sf1 <- sf1_score
```

```{r, fig.width = 3, fig.height = 3}
data.frame(
    "UMAP1" = neurons@reductions$umap@cell.embeddings[,1],
    "UMAP2" = neurons@reductions$umap@cell.embeddings[,2],
    "Sf1" = sf1_score
    ) %>%
    mutate(Sf1 = ifelse(Sf1 > 10, 10, ifelse(Sf1 < -10, -10, Sf1))) %>%
    ggplot(aes(x = UMAP1, y = UMAP2, color = Sf1)) +
    geom_point(stroke = 0, size = 0.5) +
    theme_void() +
    scale_color_gradient2(name = "TRAP\nloading",
                          labels = c("<-10", -5, 0, 5, ">10"),
                          mid = "gray95") +
  theme(legend.position = "top", legend.direction = "horizontal",
        legend.key.height = unit(0.03, "in"),
        legend.key.width = unit(0.15, "in"),
        legend.text = element_text(size = 6),
        legend.title = element_text(size = 6))
```


This is also consistent with cells that have the most the *Fezf* and *Nr5a1* expression.

## Calling VMH populations
To do this in an unbiased way, I'll call VMH neurons as clusters that are VGLUT2 (*Slc17a6*)-positive, *Gad1*-negative, either *Fezf1*- or *Nr5a1*-positive, and have a positive *Nr5a1* TRAP score.

```{r}
marker_genes <- c("SLC17A6", "GAD1", "FEZF1", "NR5A1")
if (!all(marker_genes %in% rownames(neurons[["RNA"]]@scale.data))) {
  neurons <- scale_data(
    neurons,
    genes = union(rownames(neurons[["RNA"]]@scale.data), marker_genes),
    groups = "Sample"
  )
}

marker_means <- cluster_means(neurons, c("SLC17A6", "GAD1", "FEZF1", "NR5A1"))
vmh_scores <- map_dbl(
  sort(unique(neurons@active.ident)),
  ~ median(sf1_score[neurons@active.ident == .x])
)
```

```{r}
vmh_clusters <- colnames(marker_means)[
  marker_means["SLC17A6", ] > 0 & marker_means["GAD1", ] < 0 &
  (marker_means["FEZF1", ] > 0 | marker_means["NR5A1", ] > 0) &
    vmh_scores > 0
]
```

```{r, fig.width = 7.5, fig.height = 0.9}
marker_means["TRAP", ] <- vmh_scores

marker_means %>% t() %>% as.data.frame() %>%
  rownames_to_column("cluster") %>%
  mutate_if(is.numeric, ~ .x > 0) %>%
  rename("Glutamatergic" = SLC17A6) %>%
  mutate(GAD1 = !GAD1) %>%
  rename("not GABAergic" = GAD1) %>%
  gather(-cluster, key = "measure", value = "result") %>%
  mutate(measure = factor(measure, levels = c(
    "NR5A1", "FEZF1", "Glutamatergic", "not GABAergic", "TRAP"
  ))) %>%
  mutate(cluster = factor(cluster, levels = colnames(marker_means))) %>%
  ggplot(aes(x = cluster, y = fct_rev(measure), fill = result)) +
  geom_tile(color = "gray20", show.legend = FALSE) +
  scale_fill_manual(values = c("gray90", "blue4")) +
  theme_void() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y = element_text(hjust = 1, vjust = 0.5))
```

This correponds to the UMAP clusters of:

```{r, fig.width = 2.5, fig.height = 2.5}
data.frame(
  "VMH" = neurons@active.ident %in% vmh_clusters,
  "UMAP1" = neurons@reductions$umap@cell.embeddings[,1],
  "UMAP2" = neurons@reductions$umap@cell.embeddings[,2]
) %>%
  ggplot(aes(x = UMAP1, y = UMAP2, color = VMH)) +
  geom_point(stroke = 0, size = 1, show.legend = FALSE) +
  scale_color_manual(values = c("gray90", "blue4")) +
  theme_void()
```

This confirms that clusters `r paste(vmh_clusters, collapse = ", ")` represent the VMH. That is `r format(sum(neurons@active.ident %in% vmh_clusters), big.mark = ",")` neurons, `r round(sum(neurons@active.ident %in% vmh_clusters)/length(neurons@active.ident), 2)*100`% of the total.

---

Saving the `neurons` object as `neurons.rds`.

```{r}
saveRDS(neurons, "neurons.rds")
```

```{r, include = FALSE}
vmh <- subset(neurons, idents = vmh_clusters)
rm(marker_means, neurons, sf1, cluster_colors, cluster_order, sf1_score,
   vmh_clusters, vmh_scores)
gc()
```


\pagebreak

# Removing outlier cells

Some cells that clustered with VMH populations are not part of the broader VMH in UMAP space. These are likely mis-clustering contaminants. I'll remove them using `DBSCAN` before proceeding with subclustering VMH.

```{r, fig.width = 3, fig.height = 2.5}
DimPlot(vmh) +
  scale_color_manual(values = vmh@misc$cluster_colors) +
  theme_void()
```

```{r}
neuron_umap <- data.frame(
  "UMAP1" = vmh@reductions$umap@cell.embeddings[, 1],
  "UMAP2" = vmh@reductions$umap@cell.embeddings[, 2]
)

# running DBSCAN to find outliers in UMAP space
dbscan::kNNdistplot(neuron_umap)
title("Choosing episilon")
abline(a = 0.35, b = 0)

clusters <- dbscan::dbscan(neuron_umap, eps = 0.35)
```

`DBSCAN` finds `r length(unique(clusters$cluster)[unique(clusters$cluster) != 0])` clusters, not counting 1 outlier cluster (in dark green):

```{r, fig.width = 2, fig.height = 2}
neuron_umap %>%
  mutate(k = clusters$cluster) %>%
  ggplot(aes(x = UMAP1, y = UMAP2, color = factor(k))) +
  geom_point(show.legend = FALSE, stroke = 0, size = 0.8) +
  scale_color_manual(values = wesanderson::wes_palette("Chevalier1")) +
  theme_void()
```

`r sum(clusters$cluster[clusters$cluster == 1])` of the neurons fall into the largest cluster (`r round(sum(clusters$cluster[clusters$cluster == 1])/length(clusters$cluster)*100, 1)`% of the total). I'll just keep these for future processing.

I'll save the outlier cells as `vmh_outliers.rds` in case I want them in the future.

```{r}
inliers <- Cells(vmh)[clusters$cluster == 1]
outliers <- Cells(vmh)[clusters$cluster != 1]
saveRDS(subset(vmh, cells = outliers), "vmh_outliers.rds")
vmh <- subset(vmh, cells = inliers)
```


\pagebreak

# Preprocessing VMH neurons

There are `r format(ncol(vmh), big.mark = ",")` VMH neurons. I'll process them separately now to identify neuronal populations.

Pipeline:
1. Finding variable genes with `Seurat`
2. Running `sctransform` with Run as a batch variable.
2. Scaling variable genes (in both RNA and SCT assays) for each sample independently
3. Running PCA on scaled data
4. Choosing PCs from the elbow on the scree plot
5. Projecting into UMAP space

```{r, include = FALSE}
vmh <- FindVariableFeatures(vmh)
vmh <- SCTransform(vmh, conserve.memory = TRUE)
var_genes <- intersect(
  vmh@assays$RNA@var.features, vmh@assays$SCT@var.features
)

# compare RNA and SCT assays for batch effects
DefaultAssay(vmh) <- "RNA"
vmh <- scale_data(vmh, genes = var_genes, groups = "Sample")
vmh <- RunPCA(vmh, npcs = 100)
vmh@reductions$pca@misc$sig_pcs <- knee_test(vmh)
vmh <- RunUMAP(vmh, dims = 1:vmh@reductions$pca@misc$sig_pcs)

umap_object <- vmh[["umap"]]
pca_object <- vmh[["pca"]]

# compare RNA and SCT assays for batch effects
DefaultAssay(vmh) <- "SCT"
vmh <- RunPCA(vmh, npcs = 100)
vmh@reductions$pca@misc$sig_pcs <- knee_test(vmh)
vmh <- RunUMAP(vmh, dims = 1:vmh@reductions$pca@misc$sig_pcs)

gc()
```

Results: Using `r length(neurons@assays$RNA@var.features)` variable genes from `Seurat` found `r length(neurons@assays$SCT@var.features)` from `sctransform`.  The `Seurat` variable features resulted in `r pca_object@misc$sig_pcs` PCs, and the `sctransform` variable genes resulted in `r neurons@reductions$pca@misc$sig_pcs` PCs.

```{r, fig.width = 7, fig.height = 3.2}
plot_umap_samples <- function(umap) {
  df <- bind_cols(as.data.frame(umap), vmh$Sample)
  ggplot(df, aes(x = UMAP_1, y = UMAP_2, color = `...3`)) +
    geom_point(stroke = 0, size = 1) +
    scale_color_manual(values = vmh@misc$sample_colors, name = NULL) +
    theme_void() +
    theme(legend.text = element_text(size = 7),
          plot.title = element_text(size = 10, hjust = 0.5))
}

cowplot::plot_grid(plotlist = list(
  plot_umap_samples(umap_object@cell.embeddings) + ggtitle("RNA"),
  plot_umap_samples(vmh@reductions$umap@cell.embeddings) + ggtitle("SCT")
))
```

The **RNA** assay looks better, cells are more "clumped", even across samples, making it more likely to find clusters.

```{r, include = FALSE}
DefaultAssay(vmh) <- "RNA"
vmh[["pca"]] <- pca_object
vmh[["umap"]] <- umap_object
rm(umap_object, pca_object)
gc()
```

\pagebreak

# Clustering VMH neurons

To cluster, I'll start with standard Louvain clustering from the kNN graph. With `k` set at 20, I'll vary the `resolution` parameter to optimize mean silhouette width.

```{r}
resolutions <- seq(0.2, 2, by = 0.2)
print(paste("Finding clusters for resolutions", paste(resolutions, collapse = ", "), "..."))
```

```{r, include = FALSE}
sig_pcs <- vmh@reductions$pca@misc$sig_pcs
vmh <- FindNeighbors(vmh, dims = 1:sig_pcs)

clusters <- map_dfc(
  resolutions,
  ~ FindClusters(vmh, resolution = .x)@active.ident
)
gc()
# get dissimilarity matrix
distances <- cluster::daisy(
  vmh@reductions$pca@cell.embeddings[, 1:sig_pcs]
)
gc()

# calculate silhouette width for each clustering
mean_widths <- apply(clusters, 2, function(x) {
  silhouette_object <- cluster::silhouette(as.numeric(x), distances)
  summary(silhouette_object)[["avg.width"]]
})

# choose best resolution based on max mean silhoutte width
best_resolution <- resolutions[mean_widths == max(mean_widths)]

# add clustering to official ident slot
vmh@active.ident <- clusters[, resolutions == best_resolution][[1]]

# remove unnecessary object
rm(resolutions, clusters, distances, mean_widths); gc()
```

Using resolution `r best_resolution`, there are `r length(unique(vmh@active.ident))` populations.

\pagebreak


# VMH neuron markers
I'll now order the clusters based on similarity, using variable genes.

```{r}
vmh@misc$cellex <- cellex(
  as.data.frame(as.matrix(vmh@assays$RNA@counts)),
  vmh@active.ident
)
vmh@misc$cellex <- vmh@misc$cellex[, levels(vmh@active.ident)]

vmh <- scale_data(vmh, group = "Sample")
cluster_order <- order_clusters(vmh, rownames(vmh@misc$cellex))

vmh@active.ident <- factor(
  vmh@active.ident,
  levels = cluster_order,
  labels = 1:length(cluster_order)
)
vmh@misc$cellex <- vmh@misc$cellex[, cluster_order]
colnames(vmh@misc$cellex) <- 1:length(cluster_order)
```

```{r}
violin_plot(vmh, c("QRFPR", "ESR1", "NFIB", "FOXP2", "LEPR", "DLK1", "FEZF1"))

cluster_names <- c(
  `5` = "QRFPR",
  `1` = "ESR1_1",
  `4` = "ESR1_2",
  `3` = "NFIB_1",
  `7` = "NFIB_2",
  `2` = "FOXP2_2",
  `6` = "LEPR"
)

vmh@active.ident <- recode(vmh@active.ident, !!!cluster_names)
vmh@active.ident <- factor(vmh@active.ident, levels = cluster_names)
vmh@misc$cellex <- vmh@misc$cellex[, names(cluster_names)]
colnames(vmh@misc$cellex) <- cluster_names
```

\pagebreak

# Comparison with mouse
```{r}
mouse <- readRDS("../mouse/vmh.rds")
```
Which clusters have highly correlated gene expression profiles across datasets?

```{r, fig.height = 2, fig.width = 6, include = FALSE}
mouse_genes <- rownames(mouse@misc$cellex)
macaque_genes <- rownames(vmh@misc$cellex)
macaque_genes <- orthologs$`Gene name`[match(macaque_genes, orthologs$`Gene name_1`)]
var_genes <- union(mouse_genes, macaque_genes)

macaque_mtx <- vmh[["RNA"]]@scale.data
mouse_mtx <- mouse[["RNA"]]@scale.data

rownames(macaque_mtx) <- orthologs$`Gene name`[
  match(rownames(macaque_mtx), orthologs$`Gene name_1`)
]

# keep genes in rownames
var_genes <- var_genes[var_genes %in% rownames(macaque_mtx)]
var_genes <- var_genes[var_genes %in% rownames(mouse_mtx)]

# Remove var genes with NA scale data
var_genes <- var_genes[!apply(macaque_mtx[var_genes, ], 1, function(x) any(is.na(x)))]
var_genes <- var_genes[!apply(mouse_mtx[var_genes, ], 1, function(x) any(is.na(x)))]

correlation <- expand.grid(levels(mouse@active.ident), levels(vmh@active.ident))
correlation$corr <- apply(correlation, 1, function(x) {
  cor(
    rowMeans(mouse_mtx[var_genes, mouse@active.ident == x["Var1"]]),
    rowMeans(macaque_mtx[var_genes, vmh@active.ident == x["Var2"]])
    )
})

correlation %>%
  ggplot(aes(x = Var1, y = fct_rev(Var2), fill = corr)) +
  geom_tile() +
  scale_fill_gradient2() +
  theme_void() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y = element_text(hjust = 1, vjust = 0.5))

write.csv(correlation, "cluster_correlation.csv", row.names = FALSE)
```

```{r}
# assign colors based on correlation with our clusters
broad_corr <- pivot_wider(correlation, names_from = "Var1", values_from = "corr")
broad_corr <- as.data.frame(broad_corr) %>% column_to_rownames("Var2")
broad_corr <- apply(broad_corr, c(1,2), function(x) ifelse(x < 0, 0, x))
broad_corr <- broad_corr[, levels(mouse@active.ident)]

# grab our colors
rgb_colors <- col2rgb(mouse@misc$cluster_colors)
rgb_colors <- rgb_colors[, levels(mouse@active.ident)]

rgb2hex <- function(v) rgb(v[1], v[2], v[3], maxColorValue = 255)
cluster_colors <- apply(broad_corr, 1, function(x) {
  total = sum(x)
  rgb_color = sapply(1:length(x), function(i) x[i]/total * rgb_colors[, i])
  rgb2hex(rowSums(rgb_color))
})

cluster_colors <- cluster_colors[
  order(factor(names(cluster_colors), levels = levels(vmh@active.ident)))
]

vmh@misc$cluster_colors <- cluster_colors
```

---

Saving the `vmh` object as `vmh.rds`.

```{r}
saveRDS(vmh, "vmh.rds")
```

\pagebreak

# Session info
```{r}
sessionInfo()
```
